python/tk_multi_workfiles/file_item.py

```diff
@@ -213,6 +220,73 @@ def _set_versions(self, value):
         self._versions = value
     versions=property(_get_versions, _set_versions)
 
+    def generate_badge(self):
+        self._badge = None
+        app = sgtk.platform.current_bundle()
+        if self._is_local:
+            # We're a work file - get the badge for the work file from the hook:
+            try:
+                self._badge = app.execute_hook_method(
+                    "hook_get_badge",
+                    "get_work_file_badge",
+                    work_file_details=self._details,
+                    work_file_path=self._path
+                )
+            except Exception:
+                # Capture exceptions raised here and log them, so as not to break
+                # the app if the hook fails.
+                app.logger.warning(
+                    "Exception raised when getting badge for work file at %s with details %s" % (self._path, self._details),
+                    exc_info=True
+                )
+        elif self._is_published:
+            # We're a publish - get the badge for the publish from the hook:
+            try:
+                self._badge = app.execute_hook_method(
+                    "hook_get_badge",
+                    "get_publish_badge",
+                    publish_details=self._publish_details,
+                    publish_path=self._publish_path
+                )
+            except Exception:
+                # Capture exceptions raised here and log them, so as not to break
+                # the app if the hook fails.
+                app.logger.warning(
+                    "Exception raised when getting badge for publish at %s with details %s" % (self._publish_path, self._publish_details),
+                    exc_info=True
+                )
+
+        if isinstance(self._badge, QtGui.QColor):
+            # If the hook returned a QColor, we'll create a dot badge of that color.
+            try:
+                self._badge = app.execute_hook_method(
+                    "hook_get_badge",
+                    "generate_badge_pixmap",
+                    badge_color=self._badge
+                )
+            except Exception:
+                # Capture exceptions raised here and log them, so as not to break
+                # the app if the hook fails.
+                app.logger.warning(
+                    "Exception raised in hook while executing generate_badge_pixmap.",
+                    exc_info=True
+                )
+
+    def _get_badge(self):
+        """
+        :returns:   The QPixmap to be used as a badge when displaying this file
+                    to the user.
+        """
+        return self._badge
+
+    def _set_badge(self, value):
+        """
+        :param value:   The QPixmap to be used as a badge when displaying this file
+                        to the user.
+        """
+        self._badge = value
+    badge = property(_get_badge, _set_badge)
```

Comment:
Using the property method is now an anti-pattern. You can simply use the `@property` and `@badge.setter` decorators. I think this was required back in the Python 2.5 because property setters didn't exist.

As long as your class is a new-style class, which `FileItem` is because it derives from `object`, `@badge.setter` will work.

Reply:
✅